microprofile/telemetry/src/main/java/io/helidon/microprofile/telemetry/HelidonTelemetryContainerFilter.java

```diff
@@ -0,0 +1,186 @@
+
+        String annotatedPath = requestContext.getUriInfo().getPath();
+        if (resourceInfo.getResourceMethod().getAnnotation(Path.class) != null) {
+            annotatedPath = resourceInfo.getResourceMethod().getAnnotation(Path.class).value();
+        }
+
+        Span span = tracer.spanBuilder(annotatedPath)
+                .setParent(parentContext)
+                .setSpanKind(SpanKind.SERVER)
+                .setAttribute(HTTP_STATUS_CODE, HTTP_OK)
+                .setAttribute(HTTP_METHOD, requestContext.getMethod())
+                .setAttribute(HTTP_SCHEME, requestContext.getUriInfo().getRequestUri().getScheme())
+                .setAttribute(HTTP_TARGET, resolveTarget(requestContext))
+                .startSpan();
+
+        Scope scope = span.makeCurrent();
+        requestContext.setProperty(SPAN, span);
+        requestContext.setProperty(SPAN_CONTEXT, Context.current());
+        requestContext.setProperty(SPAN_SCOPE, scope);
+
+
+        // Handle OpenTelemetry Baggage from the current OpenTelemetry Context.
+        handleBaggage(requestContext, Context.current());
+
+    }
+
+    @Override
+    public void filter(final ContainerRequestContext request, final ContainerResponseContext response) {
+
+        if (LOGGER.isLoggable(System.Logger.Level.TRACE)) {
+            LOGGER.log(System.Logger.Level.TRACE, "Closing Span in a Container Request");
+        }
+
+        Context context = (Context) request.getProperty(SPAN_CONTEXT);
+        if (context == null) {
+            return;
+        }
+
+        try {
+            Span span = (Span) request.getProperty(SPAN);
+            span.setAttribute(HTTP_STATUS_CODE, response.getStatus());
+            span.end();
+
+            Scope scope = (Scope) request.getProperty(SPAN_SCOPE);
+            scope.close();
+        } finally {
+            request.removeProperty(REST_RESOURCE_CLASS);
+            request.removeProperty(REST_RESOURCE_METHOD);
+            request.removeProperty(SPAN);
+            request.removeProperty(SPAN_SCOPE);
+        }
+    }
+
+    // Resolve target string.
+    private String resolveTarget(ContainerRequestContext requestContext){
+        String path = requestContext.getUriInfo().getRequestUri().getPath();
+        String rawQuery = requestContext.getUriInfo().getRequestUri().getRawQuery();
+        if (rawQuery != null && !rawQuery.isEmpty()) {
+            return path + "?" + rawQuery;
+        }
+        return path;
+    }
+
+    // Extract OpenTelemetry Baggage from Request Headers.
+    private void handleBaggage(ContainerRequestContext containerRequestContext, Context context){
+        List<String> baggageProperties = containerRequestContext.getHeaders().get("baggage");
+        if (baggageProperties != null) {
+            for (String b : baggageProperties) {
+                String[] split = b.split("=");
+                if (split.length == 2) {
```

Comment:
What if it is not 2? Right now it is ignored. Is this correct behavior?

Reply:
They are provided as pairs. This should be sufficient.